### PR TITLE
[SPARK-41049][SQL] Revisit stateful expression handling

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

Spark has a `Stateful` trait for stateful expressions. The basic idea is to have fresh copies of stateful expressions before evaluating them. This is to avoid issues caused by the flexible DataFrame APIs:
1. A single expression instance may appear more than once in the expression tree. We have to replace it with fresh copies to avoid sharing states.
2. An expression tree can be evaluated by multiple callers at the same time. We have to use fresh copies before expression evaluation to avoid sharing states.

However, the handling of stateful expression has several problems. This PR fixes all of them:
1. We should use fresh copies with codegen as well. If the root expression extends `CodegenFallback`, then the expression tree will be evaluated using the interpreted mode, even with the codegen code path.
2. The fresh copies will be dropped if the stateful expression is deeply nested (3 layers).
3. `InterpretedSafeProjection` never implemented initialize() for initializing Nondeterministic expressions.
4. `ConvertToLocalRelation` called a `InterpretedMutableProjection` constructor which did not implement the existing Stateful-copying logic. I fixed this by moving that logic out of a factory method and into class's main constructor, guaranteeing that it will always run.
5. Stateful expression is not always nondeterministic, e.g. `ScalaUDF`. I removed the `Stateful` trait and added a `def stateful: Boolean` function in `Expression`.

### Why are the changes needed?

Fix stateful expression handling

### Does this PR introduce _any_ user-facing change?

Yes, now we never share states for stateful expressions, which may produce wrong result.

### How was this patch tested?

new tests

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala [hunk 1]
```diff
@@ -36,18 +36,12 @@ class InterpretedMutableProjection(expressions: Seq[Expression]) extends Mutable
...
-  } else {
-    expressions
-  }
+  private[this] val exprs = prepareExpressions(expressions, subExprEliminationEnabled)
 
   private[this] val buffer = new Array[Any](expressions.size)
 
   override def initialize(partitionIndex: Int): Unit = {
-    expressions.foreach(_.foreach {
+    exprs.foreach(_.foreach {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala:44]: This is a pre-existing bug. The final expressions we use is `exprs`, not `expressions`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala [hunk 2]
```diff
@@ -117,10 +111,6 @@ object InterpretedMutableProjection {
    * Returns a [[MutableProjection]] for given sequence of bound Expressions.
    */
   def createProjection(exprs: Seq[Expression]): MutableProjection = {
-    // We need to make sure that we do not reuse stateful expressions.
-    val cleanedExpressions = exprs.map(_.transform {
-      case s: Stateful => s.freshCopy()
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala:122]: The old usage of `.transform` here contained a subtle bug related to how `fastEquals` works.

Let's say that we have a tree which looks like this:

> Outer(Middle(Stateful()))

where `Outer` and `Middle` are non-Stateful expressions.

When the `.transform` is applied to `Stateful()` and `.freshCopy()` is called, the returned value will be `==` to the original Stateful expression but will have a different object identity (because it's a fresh object). Internally, `.transform` will use `fastEquals` to check whether the transformation modified the node. `Stateful` overrides `fastEquals` so that it only considers object identity, so the transform will return the `freshCopy()` result.

At the next level up, Middle will check whether any of its children have been changed in the recursive bottom-up transformation (see [childrenFastEquals() in withNewChildren()](https://github.com/apache/spark/blob/9305cc744d27daa6a746d3eb30e7639c63329072/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala#L455-L456), which is called from `mapChildren()`). It will detect that its children have changed, so the transform will return a new Middle node.

Finally, at the top level, `Outer` will perform the same check to see if any of its children have changed. This time, however, it will be calling `Middle.fastEquals` instead of `Stateful.fastEquals`. Middle's fastEquals method is the regular implementation which also considers object equality. Both the original and new `Middle` nodes will be `==`, so `fastEquals` will be true and `Outer` will conclude that its children have not been changed by the transformation and the original `Outer` will be returned (losing the copy of the stateful expression).

In other words, the old `.transform` and copying logic here was incorrect if the Stateful expression was nested more than a single level deep.

In this PR I chose to fix this by adding a `freshCopyIfContainsStatefulExpression()` method to `Expression` which implements a custom tree traversal considers _only_ object identity when determining whether the transform has changed a node or a node's children.

**Comment by cloud-fan** [comment]
cc @viirya @HyukjinKwon @gengliangwang @allisonwang-db

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  protected lazy val runtime =
+    new SubExprEvaluationRuntime(SQLConf.get.subexpressionEliminationCacheMaxEntries)
+
+  protected def prepareExpressions(
+      exprs: Seq[Expression],
+      subExprEliminationEnabled: Boolean): Seq[Expression] = {
+    // We need to make sure that we do not reuse stateful expressions.
+    val cleanedExpressions = exprs.map(_.freshCopyIfContainsStatefulExpression())
+    if (subExprEliminationEnabled) {
+      runtime.proxyExpressions(cleanedExpressions)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala:33]: Hmm, will sub-expr elimination extract common stateful expressions as common expr and break the rule (not reusing)?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      cleanedExpressions
+    }
+  }
+
+  /**
+   * Initializes internal states given the current partition index.
+   * This is used by nondeterministic expressions to set initial states.
+   * The default implementation does nothing.
+   */
+  def initialize(partitionIndex: Int): Unit = {}
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala:44]: This should be called after `prepareExpressions` is finished, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/InterpretedMutableProjection.scala:122]: Oh, this is a nice catch!

**Review by viirya — COMMENTED** [review]
The proposed logic looks good to me. There are some test failures that looks related.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 5]
```diff
@@ -127,6 +125,54 @@ abstract class Expression extends TreeNode[Expression] {
...
+        i += 1
+      }
+      res
+    }
+    // If the children contain stateful expressions and get copied, or this expression is stateful,
+    // copy this expression with the new children.
+    if (anyChildChanged || stateful) {
+      CurrentOrigin.withOrigin(origin) {
+        val res = withNewChildrenInternal(newChildren)
+        res.copyTagsFrom(this)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:168]: The keys in `tags` still refer to original tree node. Is it okay?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala (hunk 3 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala:33]: 2 stateful but deterministic expressions always produce the same result given the same input sequence. So it's OK to apply sub-expr elimination.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala:44]: yes

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:168]: How is this possible? keys are basically strings, right?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionsEvaluator.scala:44]: or more specifically, the implementation should initialize the final expression that it actually uses.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 5 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:168]: Note: the code here basically follows `TreeNode.withNewChildren`

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:168]: Yea, you're correct. It is just a string for node name.

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks for the review, merging to master!
